build: Enable tsec checks for critical packages.

uraj opened the pull request:
tsec is a static analyzer that discovers Trusted Types violations.
Deploy tsec to make sure there will be no TT regression in the core
package. Existing violations have been reviewed and exempted in
packages/tsec-exemption.json.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [x] Build related changes
- [x] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

uraj committed:
build: Enable tsec checks for critical packages.

tsec is a static analyzer that discovers Trusted Types violations.
Deploy tsec to make sure there will be no TT regression in several
critical packages, including core, platform-browser, platform-server
and their dependencies. Existing violations have been reviewed and
exempted in packages/tsec-exemption.json. Future changes to the
exemption list requires security review.

uraj commented:
@josephperrott This is the PR that enables TT static analysis for the core package. Could you take a look?

IgorMinar reviewed (commented):
Thanks @uraj for sending us this PR! I'm supportive of adding tsec to our repo.

Overall I like where this is heading, I left a few comments primarily asking for clarification on your intention here.

If we could change the PR to run these tests via bazel, then that would be better, but I can't easily estimate how much more work that would be. @josephperrott might have a better understanding. Maybe running it via the new https://bazelbuild.github.io/rules_nodejs/ that autogenerate bazel rules for npm binaries (see the babel example in that link) would make it trivial if we can collect all the right inputs easily.

Please take a look at the other comments. Thank you again.

IgorMinar commented on packages/tsec-exemption.json:
```diff
@@ -0,0 +1,25 @@
+/**
```
oh, one more thing

can you please add this file (and possibly all tsec tsconfig files) to the pullapprove security group: https://github.com/angular/angular/blob/9fb79d38aade67f6f3bcdac0ffa93b8806baa215/.pullapprove.yml#L732-L738 so that we request the reviews from the right people for changes to these files? thanks

uraj replied:
Done.

uraj commented on packages/core/tsconfig-tsec.json:
```diff
@@ -0,0 +1,20 @@
+  "extends": "../tsconfig-tsec-base.json",
+  "compilerOptions": {
+    "composite": true,
+    "outDir": ".tsec"
```
Tsetse (the checker engine used by tsec) used to be part of @bazel/typescript, but the support was discontinued due to the TS team being understaffed. I will see if I can make a custom rule to run tsec.

uraj replied:
I was able to do certain amount of bazel integration. There are several caveats though.
  - tsconfig-tsec.json files are still not auto-generated and have to be maintained manually to make it consistent with BUILD.bazel. The maintenance cost should be low though.
  -  The bazel rule `tsec_test` does not check transitive dependencies, so each dependency needs to have its own `tsec_test`. This is hard to fix, unless we can have some upstream patches for the vanilla `ts_library` bazel rule.

uraj commented on packages/localize/tsconfig-tsec.json:
```diff
@@ -0,0 +1,9 @@
+/** For building //packages/localize. */
```
I saw `//packages/localize` as one of `core`'s dependencies in packages/core/BUILD.bazel, so I though I will need to build it as a reference of `core`. But it turns out it's actually not needed.

I will add `platform-browser` and `platform-server`.

uraj commented on packages/tsconfig-tsec.json:
```diff
@@ -0,0 +1,10 @@
+    "outDir": ".tsec",
+    "types": ["hammerjs"]
+  },
+  "include": ["*.ts"]
```
You're right that this one doesn't cover anything meaning full (only empty.ts). I've changed this to a master config that builds all packages we want to check.

uraj commented on packages/zone.js/lib/tsconfig-tsec.json:
```diff
@@ -0,0 +1,10 @@
+    "outDir": ".tsec",
+    "types": ["node"]
+  },
+  "files": ["zone.ts"]
```
The original plan is to only check `core`, which only depends on `zone.d.ts`. Do we actually want to check `browser/*`? If so, I can include it as one of the to-be-checked packages.

uraj committed:
build: Turn tsec checks into bazel tests.

Introduce two new bazel rules: tsec_test and tsec_config, for
describing the tsec checks and the tsconfig file needed for such
checks, respectively. Currently, tsec_test only checks the srcs
of a ts_library or ng_module. It does not check direct or transitive
dependencies. Also, tsconfig files need to be manually maintained
to make sure tsec can read all necessary input (including global
symbols).

uraj commented:
@IgorMinar Thanks for the review! I managed to implement some basic integration for tsec and Bazel (details are in my reply to code review comments). PTAL.

IgorMinar reviewed (commented):
At a high level this looks good to me. So much better than the original version. I think it was @josephperrott who suggested that this check should really be done as part of the npm_package/ng_package rules but i don't think that that implementation is strictly needed for the initial version. 

I'd love for @devversion or @josephperrott to sign off on the bazel rule implementation, but again I think this is pretty close to a mergable V1. Thanks for all the iterations on this.

josephperrott commented on tools/tsec.bzl:
```diff
@@ -0,0 +1,143 @@
+        target: the ts_library or ng_module target to be checked
+        tsconfig: the tsec_config target used for configuring tsec
+    """
+    all_data = name + "_all_data"
```
Historically we have preferred to use python's string formatting for substitutions like this.

```suggestion
    all_data = "%s_all_data" % name
```

josephperrott replied:
Can we just do the substitutions in place, or change the variable name used?  It seems odd to have `all_data` as a string that we use as a parameter on `_all_data`

uraj replied:
Done.

josephperrott commented on tools/tsec.bzl:
```diff
@@ -0,0 +1,143 @@
+
+    return [DefaultInfo(files = depset(files))]
+
+_all_data = rule(
```
I think we can find a better name for this than `_all_data`, how about something like `_all_dependencies` or `_all_transitive_deps`

uraj replied:
Changed to `_all_transitive_deps`.

josephperrott commented on packages/compiler/BUILD.bazel:
```diff
@@ -12,6 +13,18 @@ ts_library(
+    deps = ["//packages:tsec_config"],
+)
+
+tsec_test(
```
For these `tsec_test` usages, they are going to be the only ones in each package they are found in.

Can we just use the name `tsec_test` instead of `<package name>_tsec_test`

uraj replied:
Done.

devversion commented on tools/defaults.bzl:
```diff
@@ -181,6 +192,16 @@ def ng_module(name, tsconfig = None, entry_point = None, testonly = False, deps
+        deps = deps,
+        tags = ["tsec"],
+    )
+
```
It seems like we can drop all of this additional targets by just using an aspect that does not have propagation enabled. The aspect will only check the `target` passed to the `tsec_test` macro and it will augment the target with the `TsLibInfo` provider. This has the following benefits:

* We keep all the tsec logic local to the `tsec.bzl` file. No other infrastructure setup is involved.
* We do not need to setup additional targets for all `ng_module` or `ts_library` targets (most of these would be unused as we only test a subset with tsec).

e.g.

```bzl

def _capture_tsec_deps_aspect_impl(target, ctx):
    """Forward `srcs` and `deps` of `ts_library` and `ng_module` macros to `_tsec_test`."""
    return [TsLibInfo(srcs = ctx.rule.attr.srcs, deps = ctx.rule.attr.deps)]

capture_tsec_deps_aspect = aspect(
    implementation = _capture_tsec_deps_aspect_impl,
)
```

and this aspect would be applied to e.g.

```bzl
_all_data = rule(
    implementation = _all_data_impl,
    attrs = {
        "tsconfig": attr.label(),
        "target": attr.label(aspects = [capture_tsec_deps_aspect]),
    },
```

What do you think? cc. @josephperrott

uraj replied:
Yes, aspects is so much better. I have updated the code.

devversion reviewed (approved):
Overall LGTM. Unfortunately this rule is not working on Windows though (we try making most of the tooling windows-compatible). The problem is that on Windows the test inputs are not symlinked, so the `.ts` sources end up staying in the actual project directory while dependencies are in the `bazel-out`. It would require module resolution through `@bazel/runfiles` to make it work. 

It's unfortunate that this won't work on Windows, but I assume it's not necessarily blocking. cc. @josephperrott and @IgorMinar for feedback on this.

uraj committed:
build: Use aspects to gather srcs and deps for tsec_test.

This is to replace the implicitly created ts_library_forwared rules and
keep changes related to tsec/bazel integration solely in tools/tsec.bzl.

josephperrott reviewed (approved):
LGTM

Per our conversation offline, I would love for us to make this work without having to have an additional set of `tsconfig`s checked in and with windows support.  But we can do these things in follow up PRs.

uraj committed:
build: Upgrade tsec to 0.1.9

Upgrade tsec version form 0.1.8 to 0.1.9. The new version has a minor
bug fix.

devversion reviewed (approved):
LGTM. Thanks for switching to the aspect. 

We should definitely look into the Windows support. Usually doing these things in follow-ups tend to be forgotten so we should create an issue for tracking IMO.

uraj commented:
@devversion I filled google/tsec#23 to track Windows compatibility.

AndrewKushnir commented:
@josephperrott there is an indication that this PR needs g3 presubmit and it looks like only `**/tsconfig-tsec.json` files are synced (see [presubmit CL](https://critique-ng.corp.google.com/cl/396412772)). Do we need those files in g3 of should we exclude them from the sync? Adding the "cleanup" label to this PR for now.

uraj commented:
@AndrewKushnir I think they can be excluded.

AndrewKushnir commented:
@uraj thanks for the comment. Could you please update the exclude list in https://github.com/angular/angular/blob/master/.github/angular-robot.yml#L42 to reflect that (and also the file in g3 that is mentioned at https://github.com/angular/angular/blob/master/.github/angular-robot.yml#L35)? That should eliminate the need to have a presubmit for this PR.

uraj committed:
build: Exclude tsconfig-tsec.json files from sync with g3

Those files are for tsec_test only and therefore irrelevant to google3.

uraj commented:
@AndrewKushnir Updated.

AndrewKushnir commented:
Thanks for the update @uraj. I've updated g3 status for this PR and removed the "cleanup" label.

AndrewKushnir commented:
Merge-assistance: it's ok to proceed with the current set of reviews, since its just testing changes around everything and not changing the code in those areas.
